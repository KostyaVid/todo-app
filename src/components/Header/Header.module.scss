.header {
  width: 100%;
  flex: 0 0 300px;
  background-image: url(./../../../public/images/bg-desktop-light.jpg);
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
  @media (max-width: 590px) {
    flex: 0 0 200px;
  }
}

.lightThemeHeader {
  background-image: url(./../../../public/images/bg-desktop-light.jpg);
}

.darkThemeHeader {
  background-image: url(./../../../public/images/bg-desktop-dark.jpg);
}

.flexRow {
  margin-top: 78px;
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: space-between;
  position: relative;
  @media (max-width: 590px) {
    margin-top: 48px;
  }
}

.title {
  color: #fffffd;
  font-size: 38px;
  font-weight: 700;
  letter-spacing: 16px;
  @media (max-width: 590px) {
    font-size: 30px;
    font-weight: 700;
    letter-spacing: 12px;
  }
}

.toggleTheme {
  width: 30px;
  height: 30px;
  position: relative;
  top: -3px;
  background-color: inherit;
  background-image: url(./../../../public/images/icon-moon.svg);
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
  @media (max-width: 590px) {
    width: 25px;
    height: 25px;
  }
}

.lightThemeToggle {
  background-image: url(./../../../public/images/icon-moon.svg);
}

.darkThemeToggle {
  background-image: url(./../../../public/images/icon-sun.svg);
}
