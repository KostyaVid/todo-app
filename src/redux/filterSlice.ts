import { createSlice, PayloadAction } from '@reduxjs/toolkit';

export interface IFilterState {
  filter: 'all' | 'active' | 'completed';
}

const initialState: IFilterState = {
  filter: 'all',
};

export const filterSlice = createSlice({
  name: 'filter',
  initialState,
  reducers: {
    setFilter: (state: IFilterState, action: PayloadAction<IFilterState>) => {
      state.filter = action.payload.filter;
    },
  },
});

export const { setFilter } = filterSlice.actions;
export default filterSlice.reducer;
