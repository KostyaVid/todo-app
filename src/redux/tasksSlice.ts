import { createSlice } from '@reduxjs/toolkit';
import type { PayloadAction } from '@reduxjs/toolkit';

export interface ITask {
  id: number;
  description: string;
  isCompleted: boolean;
  onClose: boolean;
  isHide: boolean;
}

export interface ITaskDrag {
  task: ITask;
  offsetX: number;
  offsetY: number;
}

export interface ITasksState {
  tasks: Array<ITask>;
  tasksLeft: number;
  lastID: number;
  taskDrag: number | null;
}

const initialState: ITasksState = {
  tasks: [
    {
      id: 0,
      description: 'Complete online JavaScript course',
      isCompleted: true,
      onClose: false,
      isHide: false,
    },
    {
      id: 1,
      description: 'Jog around the park 3x',
      isCompleted: false,
      onClose: false,
      isHide: false,
    },
    {
      id: 2,
      description: '10 minutes meditation',
      isCompleted: false,
      onClose: false,
      isHide: false,
    },
    {
      id: 3,
      description: 'Road for 1 hour',
      isCompleted: false,
      onClose: false,
      isHide: false,
    },
    {
      id: 4,
      description: 'Pick up groceries',
      isCompleted: false,
      onClose: false,
      isHide: false,
    },
    {
      id: 5,
      description: 'Complete Todo App on Frontend Mentor',
      isCompleted: false,
      onClose: false,
      isHide: false,
    },
  ],
  tasksLeft: 5,
  lastID: 5,
  taskDrag: null,
};

export const tasksSlice = createSlice({
  name: 'tasks',
  initialState,
  reducers: {
    addTask: (state: ITasksState, action: PayloadAction<string>) => {
      state.lastID++;
      state.tasksLeft++;
      let task: ITask = {
        id: state.lastID,
        description: action.payload,
        isCompleted: false,
        onClose: false,
        isHide: false,
      };
      state.tasks.push(task);
    },

    clearCompletedTasks: (state: ITasksState) => {
      let j = state.tasks.length;
      for (let i = 0; i < j; i++) {
        if (state.tasks[i].isCompleted) {
          state.tasks.splice(i, 1);
          j--;
          i--;
        }
      }
    },
    toggleCompleteByID: (state: ITasksState, action: PayloadAction<number>) => {
      let posInArray = 0;
      state.tasks.some((task: ITask, index: number) => {
        if (task.id === action.payload) {
          posInArray = index;
          return true;
        }
        return false;
      });

      state.tasks[posInArray].isCompleted = !state.tasks[posInArray].isCompleted;
      state.tasks[posInArray].isCompleted ? (state.tasksLeft -= 1) : (state.tasksLeft += 1);
    },
    deleteTaskByID: (state: ITasksState, action: PayloadAction<number>) => {
      let posInArray = 0;
      state.tasks.some((task: ITask, index: number) => {
        if (task.id === action.payload) {
          posInArray = index;
          return true;
        }
        return false;
      });
      if (!state.tasks[posInArray].isCompleted) state.tasksLeft -= 1;
      state.tasks[posInArray].onClose = true;
    },

    deleteAndRemoveTaskByID: (state: ITasksState, action: PayloadAction<number>) => {
      let posInArray = 0;
      state.tasks.some((task: ITask, index: number) => {
        if (task.id === action.payload) {
          posInArray = index;
          return true;
        }
        return false;
      });
      state.tasks.splice(posInArray, 1);
    },
    setDragID: (state: ITasksState, action: PayloadAction<number | null>) => {
      state.taskDrag = action.payload;
    },
    toggleTasksByDrag: (state: ITasksState, action: PayloadAction<number>) => {
      if (state.taskDrag != null) {
        let tempTask: ITask | undefined = undefined;
        let tempTaskID = 0;
        for (let j = 0; j < state.tasks.length; j++) {
          if (state.tasks[j].id === state.taskDrag) {
            tempTask = { ...state.tasks[j] };
            tempTaskID = j;
            break;
          }
        }
        if (tempTask) {
          for (let i = 0; i < state.tasks.length; i++) {
            if (state.tasks[i].id === action.payload) {
              state.tasks.splice(tempTaskID, 1);
              state.tasks.splice(i, 0, tempTask);
              return;
            }
          }
        }
      }
    },
  },
});

export const {
  addTask,
  deleteTaskByID,
  toggleCompleteByID,
  clearCompletedTasks,
  deleteAndRemoveTaskByID,
  setDragID,
  toggleTasksByDrag,
} = tasksSlice.actions;
export default tasksSlice.reducer;
