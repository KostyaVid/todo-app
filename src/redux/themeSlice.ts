import { createSlice } from '@reduxjs/toolkit';
import type { PayloadAction } from '@reduxjs/toolkit';

export interface ThemeState {
  theme: string;
}

const initialState: ThemeState = {
  theme: 'light',
};

export const themeSlice = createSlice({
  name: 'theme',
  initialState,
  reducers: {
    setTheme: (state: any, action: PayloadAction<string>) => {
      state.theme = action.payload;
    },
    toggleTheme: (state: any) => {
      if (state.theme === 'light') {
        state.theme = 'dark';
      } else {
        state.theme = 'light';
      }
    },
  },
});

// Action creators are generated for each case reducer function
export const { setTheme, toggleTheme } = themeSlice.actions;

export default themeSlice.reducer;
