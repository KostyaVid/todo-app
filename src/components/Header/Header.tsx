import { useDispatch, useSelector } from 'react-redux';
import { toggleTheme } from '../../redux/themeSlice';
import Container from '../Container/Container';
import s from './Header.module.scss';

const Header = () => {
  const theme: string = useSelector((state: any) => state.theme.theme);
  const dispatch = useDispatch();

  return (
    <header
      className={[s.header, theme === 'light' ? s.lightThemeHeader : s.darkThemeHeader].join(' ')}>
      <Container>
        <div className={s.flexRow}>
          <h1 className={s.title}>TODO</h1>
          <button
            className={[
              s.toggleTheme,
              theme === 'light' ? s.lightThemeToggle : s.darkThemeToggle,
            ].join(' ')}
            onClick={() => {
              dispatch(toggleTheme());
              document.body.classList.toggle('lightTheme');
              document.body.classList.toggle('darkTheme');
            }}></button>
        </div>
      </Container>
    </header>
  );
};

export default Header;
