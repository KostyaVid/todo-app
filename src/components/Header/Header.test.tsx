import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { store } from '../../redux/store';
import Header from './Header';

test('render header', () => {
  const component = renderer.create(
    <Provider store={store}>
      <Header />
    </Provider>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
