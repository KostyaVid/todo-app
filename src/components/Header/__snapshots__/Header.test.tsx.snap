// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render header 1`] = `
<header
  className="header lightThemeHeader"
>
  <div
    className="container"
  >
    <div
      className="flexRow"
    >
      <h1
        className="title"
      >
        TODO
      </h1>
      <button
        className="toggleTheme lightThemeToggle"
        onClick={[Function]}
      />
    </div>
  </div>
</header>
`;
